File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    #속한 노래 많이 재생 장르 먼저
    #장르 내에서 많이 재생된 노래 먼저
    #장르 내 재생횟수 같은 노래 중 고유 번호 낮은 순
    
    #딕셔너리 하나 해서 전체 장르 별 재생횟수 내림차순 정렬
    #genres 리스트 순회하면서 정렬된 장르마다 리스트 만들고 재생횟수 번호 별 리스트 만들고 정렬
    #그리고 상위 두개 뽑기
    
    playGenreDict = {} #장르 별 재생횟수 저장 딕셔너리
    for i in range(len(genres)): #장르 별 재생횟수 저장하기
        playGenreDict[genres[i]] = playGenreDict.get(genres[i], 0) + plays[i]
    
    genreList = sorted(playGenreDict.items(), key = lambda x : x[1], reverse = True)
    
    answer = []
    
    genreIdx = 0
    while genreIdx < len(genreList):
        tmp = []
        for i in range(len(genres)):
            if genres[i] == genreList[genreIdx][0]:
                tmp.append((plays[i], i))
        tmp.sort(key = lambda x : (-x[0], x[1]))
        
        answer.append(tmp[0][1])
        if(len(tmp) >= 2):
            answer.append(tmp[1][1])
        
        genreIdx += 1
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Solution {

    public int[] solution(String[] genres, int[] plays) {

        // 장르별 총 재생 횟수
        Map<String, Integer> genreTotal = new HashMap<>();

        // 장르별 노래 목록
        // int[] = {고유번호, 재생 횟수}
        Map<String, List<int[]>> songs = new HashMap<>();

        for (int i = 0; i < genres.length; i++) {
            String genre = genres[i];

            genreTotal.put(
                genre,
                genreTotal.getOrDefault(genre, 0) + plays[i]
            );

            songs.putIfAbsent(genre, new ArrayList<>());
            songs.get(genre).add(new int[]{i, plays[i]});
        }

        // 장르를 총 재생 횟수 내림차순으로 정렬
        List<String> genreList = new ArrayList<>(genreTotal.keySet());

        genreList.sort((g1, g2) ->
            genreTotal.get(g2) - genreTotal.get(g1)
        );

        List<Integer> result = new ArrayList<>();

        // 많이 재생된 장르부터 순회
        for (String genre : genreList) {

            List<int[]> songList = songs.get(genre);

            // 장르 내부 노래 정렬
            songList.sort((a, b) -> {
                // 재생 횟수가 같으면 고유번호가 낮은 순서
                if (a[1] == b[1]) {
                    return a[0] - b[0];
                }

                // 재생 횟수가 많은 순서
                return b[1] - a[1];
            });

            // 장르별 최대 2곡 선택
            result.add(songList.get(0)[0]);

            if (songList.size() >= 2) {
                result.add(songList.get(1)[0]);
            }
        }

        int[] answer = new int[result.size()];

        for (int i = 0; i < result.size(); i++) {
            answer[i] = result.get(i);
        }

        return answer;
    }
}